Aircraft kinematics part 7: Fix aircraft attack patterns to make use of the new kinematics.

tovl opened the pull request:
This is a follow up to #18320, #18322, #18360, #18383, #18384 and #18385.

This is the final part of the aircraft kinematics overhaul aside for one significantly less technical and much more aesthetic follow up to fine-tune all the aircraft parameters in TS. That one will come later though and is not needed to assess the changes here, since some sensible first-order values are included here. When testing the PRs mentioned above, I recommend to look here to see what things will look like when everything is taken together.

This one is relatively simple compared to the previous in the series and mostly picking the fruits from what was set up earlier. It does three things: 

- Strafe attacking runs now work properly for sliding aircraft. They will now always move forward in a straight line instead of trying to immediately turn around. They will still take moving targets into account when determining the proper distance to come around for another pass. When `BodyTurnSpeed` and `TurnSpeed` are unequal, they will time their turn so that they are properly facing the target again at the moment they are back in range. Contrary to the Strafe attack pattern, In the default attack-pattern non-hovering sliders will try to face their target as much as possible, ignoring their flight direction, but they will now always properly pass their target instead of going back to where they came from.
- Aircraft with the Hover attack pattern now make use of the new hovering kinematics, meaning that they will keep some of their momentum and properly decelerate instead of jerking to a standstill. They are also able to turn and adjust their altitude at the same time.
- Added a new parameter `AttackAltitude` to `AttackAircraft` which, if defined, allows aircraft to dive towards their target when attacking, pulling up in the rebound. This makes use of the same mechanics added in #18385 for partial VTOL landing. This is applied here to make Banshees into actual dive-bombers.

tovl committed:
Add IdleSpeed to aircraft trait.

tovl committed:
Let TS aircraft turn slower when circling.

tovl committed:
Separate aircraft body facing and flight facing.

tovl committed:
Remove move overrides from Fly and FlyIdle.

tovl committed:
Convert aircraft turn speed parameters to WAngle.

tovl committed:
Add updaterule.

tovl committed:
Implement acceleration/deceleration for aircraft.

tovl committed:
Make aircraft pitch dependent on horizontal speed and direction.

tovl committed:
Add acceleration to TS aircraft.

tovl committed:
Implement angular acceleration for aircraft.

tovl committed:
Make aircraft roll dependent on actual turn speed.

tovl committed:
Simplify calls to FlyTick.

tovl committed:
Let sliding aircraft smoothly turn towards their target facing in mid-flight before landing.

tovl committed:
Separate TurnSpeed and BodyTurnSpeed for TS aircraft.

tovl committed:
Add angular acceleration to TS aircraft.

tovl committed:
Let aircraft slow down to tighten their turn circle.

tovl committed:
Implement acceleration for vertical take-off/landing.

tovl committed:
Allow VTOL landings to move diagonally

tovl committed:
Implement deceleration on landing for non-VTOLs.

tovl committed:
Add VTOL acceleration to TS aircraft.

tovl committed:
Let aircraft predict and follow terrain height.

tovl committed:
Fix glitches due to TakeOffs triggering at the wrong moment.

tovl committed:
Allow aircraft to visibly tilt when changing altitude.

tovl committed:
Add slope tilting to TS aircraft.

tovl committed:
Let aircraft turn towards their target pitch and roll in midflight.

tovl committed:
Allow aircraft to aim at a specific target altitude and let VTOLs define an altitude to initiate vertical landing.

tovl committed:
Let strafing work properly with sliding aircraft.

tovl committed:
Allow aircraft to dive when attacking.

tovl committed:
Maintain momentum on hovering attack and following.

tovl committed:
Add strafing attacks to TS aircraft.

pchote commented:
I see the potential here, but a few things stick out while testing (i'll let you work out which specific PR has the problems):
* Aircraft drop too low after descending from a cliff and then visibly bounce back up
* I'm really not a fan of the height-changing pitch effect - it interacts badly IMO with the standard pitching behavior, and overrides the hard limits set on carryalls and harpies where it is actually broken.
* The dynamics of the banshee feel wrong IMO - like a leaf being blown around in the wind rather than a massive thing throwing its weight around using strong thrust.

tovl commented:
> I'm really not a fan of the height-changing pitch effect - it interacts badly IMO with the standard pitching behavior, and overrides the hard limits set on carryalls and harpies where it is actually broken.

This is actually toggleable in yaml and disabled by default. It should have been disabled for carryalls. It should definitely be enabled for banshees as well as any non-helicopter-like airplanes. Banshees pitching forward like a helicopter also doesn't make sense.

tovl committed:
Adjust Banshee flight parameters.

tovl commented:
Added a commit tweaking the banshee parameters to make it feel a bit more aggressive and less floaty.

tovl commented:
That hovering aircraft drop too low after descending is obviously a bug in #18384. Apparently the braking distance is estimated incorrectly, so the aircraft overshoots.

penev92 commented:
Okay, I haven't looked at any of the code yet, but for the moment am going to comment on the units' feel in-game.

There are definitely cool things here, but some of them also don't come off as too realistic.
 - Orca Bomber attack pattern is awesome (well how it turns depends on the angle I guess, cause it's not consistent), but then you realize this is supposed to be a huge, heavy bomber with just a couple of fans providing trust, so that behaviour is unlikely (I think @pchote said something similar above about the Banshee?).
 - The Banshee looks cool if we agree on "anti-gravity propulsion", but even then I can make it stand on its nose with a bit of micro which is ... a bit odd.
 - Orca/Harpy turning isn't exactly polished (idk if that was a goal of these PRs or is separate) - they start turning like airplanes but then just slide sideways like helicopters (especially when ordered while moving).
 - Going up/down hills is still a bit problematic - _especially_ going down hills - actors that have pitched forward pitch even harder, but the change in altitude is barely noticeable and generally comes much later than it should.
 - The carryall has a host of issues but I saw something on that mentioned in the discussion above so not really a priority.

This is why I think it is important to establish how we want them to feel and act first and then do the actual (YAML) changes (engine support for all this can still get in as I'm sure modders can have a great time playing around with what could be available).

P.S.: While testing this combined with the proposal in #18400 I tweaked that a bit to better illustrate the altitude changes I mentioned above:
<details>
<summary>YouTube videos</summary>
Notice how the colour changes with altitude:

[![](https://img.youtube.com/vi/BtXJ6zbiHuk/0.jpg)](https://www.youtube.com/watch?v=BtXJ6zbiHuk "Orca altitude test")
[![](https://img.youtube.com/vi/SwWMtjJaYhE/0.jpg)](https://www.youtube.com/watch?v=SwWMtjJaYhE "Banshee altitude test")

</details>

penev92 commented:
By the way, all the non-support aircraft in the gen1 mods are badly broken. Interesting enough, D2k CHOAM Frigate is as well, but TD C17 isn't.
Sorry, haven't looked at which PR is at fault.

tovl commented:
Those altitude indicators look really nice! Would you mind sharing your code? It would be really helpful when debugging this.

To be absolutely clear what my intention is here: The aircraft behaviours that these PRs implement for TS are not intended to be final and polished for those individual units. The yaml changes included are intended as the minimal required to be able to demonstrate the potential of the engine additions to make those behaviours possible. 

I completely agree that it is important to discuss and agree on how we want the individual units to behave before we start doing all the yaml tweaks. I also believe it is important to establish clear boundary conditions on what the possibilities and limitations of the engine are before discussing in detail how the units should behave. Otherwise we might decide we want things that are way too hard to implement, or lean towards a compromise that may turn out to be unnecessary.

Therefore, the idea of these PRs is to expand the capabilities of the engine to support more detailed kinematic models with half and eye towards the needs of TS, but without getting too fuzzed about the details. Then we can continue the discussion started in #18314 and #18319 and that threatened to derail #18231. And when we reach a consensus there, _then_ I will file a PR which will hopefully only need yaml changes to implement the final behaviours.

About the helicopter behaviour: I used a flight model broadly based on the model used by Westwood for helicopters, which is to use two separate facings: one for the flight direction and one for the orientation of the sprite/voxel. The facings can have different turn speeds attached to them and will independently move towards the target facing. This can sometimes lead to some odd results, but generally is a simple way to get some reasonably convincing helicopter behaviour without going overboard in simulating the exact propulsion mechanisms.

The only thing that I find a bit awkward in this model, is that it is possible for one facing to go clockwise while the other goes anti-clockwise. This is what you observed with the Orca bomber attack runs. This sort of thing happened in the original too sometimes, and I think it gave the Banshee the alien character that it needs, but I don't think it makes much sense for any other aircraft.

tovl commented:
Regarding the altitude change: The height prediction might be a bit too conservative when going downhill, which I might look into. In general though, just increasing the pitch speed in yaml should make altitude changes more immediate.

As for the visual pitching: I don't think combining visual pitching on altitude change with visual pitching on speed increase combines well here either. The simple solution would be to just do one of them, but not both. This would be preferred for most aircraft anyway, but perhaps orca fighters might want to do both? In that case, we could have a hard limit on pitch that applies to the combined effect, which would probably make it look a lot better.

penev92 committed:
DUMP  Added RenderAircraftLocationIndicator trait

penev92 commented:
> Those altitude indicators look really nice! Would you mind sharing your code? It would be really helpful when debugging this.

Thanks. I've pushed it to https://github.com/penev92/OpenRA/tree/positionIndicator_for_tovl for the moment and will hopefully get around to polishing it and opening a PR for it to address #18400. There is a commented-out line for the "subtle" version.

> I completely agree that it is important to discuss and agree on how we want the individual units to behave

Agreed. Usually what we do is make the engine changes and then do the feature showcasing in a test commit at the end that would get removed after review and testing and before merging (in your case you can leave all that to the last PR to show off what the previous ones build up to). If your intention is to get the engine changes merged before opening the floodgates to YAML configuration discussions I strongly recommend you go that way!

> And when we reach a consensus there, then I will file a PR which will hopefully only need yaml changes to implement the final behaviours.

If we can make it so, then that would be great. Seems a bit optimistic, but great.

pchote commented:
My main worry is that at the end all of all this we might realize that dynamical physics simply don't fit, and that a simpler unrealistic model feels and plays better.

We don't need to get the final balance perfect, but I do think it is important to make sure that things like "they start turning like airplanes but then just slide sideways like helicopters (especially when ordered while moving)." can be fixed at all using the options that are available right now - so we can work out new options if they can't.

tovl commented:
Thanks for the code Penev! that will be really helpful.

Tbh, I don't recognize  "they start turning like airplanes but then just slide sideways like helicopters (especially when ordered while moving)."  at all here, and if they did I also don't see how that is a bad thing. Can you explain what you think the "right" behaviour would be then? A bit more constructive criticism would be helpful, instead of just: "It's different, so it's wrong". Same thing goes for the Banshee, which I have now shown two quite different variations of, and the feedback so far is "It feels weird" which is completely unhelpful.

Besides, there is a lot of flexibility in this model, _because_ it is dynamic. Just by quantitatively adjusting some parameters a little bit, you can get quite a different feel. Just compare the Banshee before and after cd22bc8 (not saying either of them is definitively how it should be, before someone puts words in my mouth). Similarly, for helicopters, independently tuning the turn speed and turn acceleration parameters can likely give you the turning behaviour you want (if only I knew what you _do_ want). If not, there is room here to add additional restraints on top of the existing model (a hard maximum difference between body and flight attitude perhaps?) or expose some more things to yaml that are currently hardcoded.

On a personal note: Frankly, I find the negativity here quite disheartening. I already gave this more of my time than I anticipated or wanted to. Once again, I feel like I have to defend myself for daring to contribute my free time to this project and have ideas of my own. Don't get me wrong: I don't want to stir up drama here and I fully expect every PR to be scrutinized, doubly so for something as involved as this. But if skepticism and apprehension  is literally _all_ the response I get after putting that much effort in, that just doesn't encourage me to continue with it. If what I'm doing here is not appreciated, just say so, and I won't bother any more.

On the other hand, if one guy complains it is too unrealistic and the other complains it is too realistic, then perhaps I'm doing something right after all :wink:

pchote commented:
I'm sorry that you interpreted my comment as a negative. I simply meant it to justify frontloading the testing here so we can be sure that we don't groupthink our way into a tricky situation! I haven't yet had time to come up with a further list of specific constructive suggestions.

tovl committed:
WIP

tovl commented:
Small update: Tweaked the flight characteristics for the Orca, Orca bomber, carryall, transport and banshee to make turning and pitching behaviour more subtle and believable. Altitude changes going downhill are now also more direct, while the visual pitching is more toned down (and completely absent for the carryall). Added a few parameters to the aircraft trait to allow better fine-tuning. I left the turning behaviour (yaw) of the harpy unchanged to provide comparison with the orca. Depending on the feedback, I intent their behaviour to be more similar than they are here.

Just to note: testing for regressions is better done by testing the related PRs in order and comments placed where the regression actually happens. Comments regarding the overall behaviour should be placed here.

teinarss commented:
Been playing around with the latest changes and comparing it vs original TS. IMO this feels very good. 👍  Less jumpy when going over slopes.

tovl committed:
Fix overshooting on hover.

tovl committed:
Fix oversight in terrain height prediction.

tovl committed:
Allow aircraft to accelerate when attacking.

tovl commented:
Another Update:
- Fixed braking distance prediction for hover mode. Aircraft tended to overshoot severely when hovering, this is no longer the case.
- Fixed an oversight in terrain height prediction that made the aircraft pull up too eagerly. This was interfering with the diving attack (among other things), which meant that it was not possible to do steep dives no matter how low `AttackAltitude` was set.
- Added an optional `AttackSpeed` parameter to `AttackAircraft` to allow aircraft to accelerate when attacking. This combines especially well with dive attacks, because they will reach top speed at exactly the same time as they reach the low point of their attack, but can be useful on it's own as well.

I believe the last two points combined make dive attacks look a lot more convincing now.

pchote commented:
Banshees are a little (or maybe a lot) too agile now: https://i.imgur.com/xKbvkUN.mp4, https://i.imgur.com/zzVPM6T.mp4

This would look pretty good if it was moving about 5x faster, but is a little weird in its current state.

tovl commented:
I forgot to mention: Currently, the terrain height and altitude change calculations assume that there is no limit to the flight pitch angle, so I removed the limits for testing. This means that banshees can in principle go straight up, which is probably not desirable. Taking into account hard pitch limits makes the logic a bit more complicated, but I plan to address that soon.

tovl committed:
Take maximumPitch into account when planning altitude change.

tovl commented:
Another update:

Terrain height prediction and altitude change calculation now take the maximum flight pitch angle into account. Included options for `maximumPitch` overrides on idling and attacking, which are used here to make banshees pitch more aggressively when dive attacking and pitch more shallow when idle circling.

teinarss commented:
Right now banshees dives a little bit to deep, is this intentional?

tovl commented:
Banshee `AttackPitch`&mdash;which overrides `MaximumPitch` when diving for attack&mdash;is currently set to 256 (unlimited), while the normal `MaximumPitch` is set to a more reasonable 128. `AttackAltitude` is also pretty low, so it is supposed to be pretty deep. Whether this is desired in the end is a matter of debate (this can look weird in combination with tall buildings), but this can all be tweaked later by adjusting said parameters. For demonstration purposes this is intended. However, it should never be able to collide with terrain while doing so (there is a hardcoded override to maintain a minimal distance to terrain at all times even when diving to prevent them from diving into a cliff that was right behind the target). If so, that would be a bug.

It is also possible that the minimal altitude reached in the dive is slightly lower than the set `AttackAltitude`, because the trajectory planning code assumes the aircraft can exactly follow the path along the minimal turn circle without taking any margin of error. This is not completely true because of the time-quantization and the order in which the variables get updated, so I would expect it is possible that the aircraft might pull up one tick too late, but this should result in a  fairly small error.

tovl commented:
FYI, I consider this feature-complete now. I've got more stuff in the pipeline, but that is not essential for evaluating the kinematics introduced here and the PR queue is long enough as it is, so I just want to get the current crop of PRs merged before going further.

Hopefully, I have by now convinced you that this is a good direction to take, so I would like to direct reviewing to the individual PRs in order starting with #18320. I will rebase and reorder some commits introduced here as soon as the previous PR is merged.

pchote commented:
Closing for now as per https://github.com/OpenRA/OpenRA/pull/18322#issuecomment-808910954.
